### PR TITLE
Add linters for existing users, orgs, and repos

### PR DESCRIPTION
**Pull request by talsafran**
This PR adds linting to check that items in the [`collections/`](https://github.com/github/explore/tree/master/collections) directory only references GitHub users, organizations, and repositories that are publicly accessible.

This means that the build will fail if _any_ item is referencing user, organization, or repository that:

- doesn't exist (or has been deleted)
- is private (or has been made private)

The cases in parentheses mean that CI will fail for a PR even if the changes in it are valid, but on GitHub some user/org/repo referenced in [`collections/`](https://github.com/github/explore/tree/master/collections) has become inaccessible because it was deleted or made private.

Tests should still pass when a user/org/repo is renamed.

### CODE REVIEW

**Comment by talsafran** [comment]
@MikeMcQuaid I'd love your 👀 here. Two things to get this ready to go:

1. In 597fed575d11c8689fdf940625c904cd39e1b134 I configured Octokit to use an OAuth token but need to set that ENV variable in Travis, which I don't have access to. I'm also not sure with which account I should create the OAuth with (I used a personal access token locally).

2. Currently we query the GitHub API once for every user/org/repo, which isn't the fastest and could cause rate limit issues. I couldn't find any way to batch API calls (unless I missed something in the [search docs?](https://developer.github.com/v3/search/#search-users)) Another alternative could be to use VCR and configure it to expire every X hours so it's still realtime-ish.

**Comment by MikeMcQuaid** [comment]
> Tests should still pass when a user/org/repo is renamed.

Would be good to add a test to verify this e.g. verify that https://github.com/mxcl/homebrew follows the redirect as expected.

> The cases in parentheses mean that CI will fail for a PR even if the changes in it are valid, but on GitHub some user/org/repo referenced in collections/ has become inaccessible because it was deleted or made private.

We could also consider adding a Travis CI daily cron job to notify us if `master` is broken.

> Currently we query the GitHub API once for every user/org/repo, which isn't the fastest and could cause rate limit issues. I couldn't find any way to batch API calls (unless I missed something in the search docs?) Another alternative could be to use VCR and configure it to expire every X hours so it's still realtime-ish. Lmk what you think.

I think it's fine to access them all. Another option rather than using the API (given we don't care about private repositories) would just be to `curl` the relevant repository and check it's not a 404.

File: Gemfile [hunk 1]
```diff
@@ -3,6 +3,7 @@ source "https://rubygems.org"
 group :test do
   gem "fastimage", "~> 2.1.0"
   gem "minitest", "~> 5.10.3"
+  gem "octokit", "~> 4.0"
```
Comment (MikeMcQuaid) [inline Gemfile]: Personally would like to see us not bother with any of the `~>` unless there's known breakages as we already get pinned versions from `Gemfile.lock`. Feel free to leave as-is if desired and I'll make a PR to remove them all.

Comment (talsafran) [reply Gemfile]: 🆒 , I opened https://github.com/github/explore/pull/202 and assigned you

**Comment by talsafran** [comment]
Ah also, one thing unanswered from my [comment above](https://github.com/github/explore/pull/198#issuecomment-370999461) – if I create an OAuth token for CI, what's a good way to do so that all maintainers have access (ie, do we have a shared account?)?

**Comment by talsafran** [comment]
🚢'ed but noticed something funky. @MikeMcQuaid am I tripping or is Rubocop just being difficult?

First it yells at me for not using `%r` for regexes ([link](https://travis-ci.org/github/explore/builds/350163682#L1513-L1517))

```
Offenses:
test/collections_test_helper.rb:18:37: C: Style/RegexpLiteral: Use // around regular expression.
VALID_USER_AND_REPO_HTTP_STATUSES = %r{\A(200|301)\z}
                                    ^^^^^^^^^^^^^^^^^
```

I fix it, and then it yells at me for not using `//` ([link](https://travis-ci.org/github/explore/builds/351569164#L832-L835))

```
test/collections_test_helper.rb:18:37: C: Style/RegexpLiteral: Use // around regular expression.
VALID_USER_AND_REPO_HTTP_STATUSES = %r{\A(200|301)\z}
                                    ^^^^^^^^^^^^^^^^^
8 files inspected, 1 offense detected
```

🤔

So hard to please...

**Comment by talsafran** [comment]
@MikeMcQuaid woops my comment was pasted wrong (updated; guess I was indeed tripping) but thanks for the helpful hint 🙌

**Pull request merged**